### PR TITLE
Py3 compat

### PR DESCRIPTION
**Pull request by ale-rt**

### CODE REVIEW

**Comment by ale-rt** [comment]
Replaces #2 after the `ale-rt:py3-compat` has been pushed to this repository.

File: src/zc/relation/README.txt [hunk 1]
```diff
@@ -107,9 +107,8 @@ name as the tokens.
     ...     # the next parts just make the tests prettier
     ...     def __repr__(self):
     ...         return '<Employee instance "' + self.name + '">'
-    ...     def __cmp__(self, other):
-    ...         # pukes if other doesn't have name
-    ...         return cmp(self.name, other.name)
+    ...     def __lt__(self, other):
+    ...         return self.name < other.name
```
Comment (mgedmin) [inline src/zc/relation/README.txt:113]: I don't like seeing only one of six rich comparison methods being defined. If you use @functools.total_ordering, you can get away by defining two of six.

File: src/zc/relation/catalog.py [hunk 2]
```diff
@@ -902,7 +903,7 @@ def yieldRelationTokenChains(self, query, relData, maxDepth, checkFilter,
         while stack:
             tokenChain, relDataIter = stack[0]
             try:
-                relToken = relDataIter.next()
+                relToken = six.next(relDataIter)
```
Comment (mgedmin) [inline src/zc/relation/catalog.py]: `next()` is a builtin since Python 2.6

File: src/zc/relation/catalog.py [hunk 3]
```diff
@@ -1158,11 +1159,11 @@ def canFind(self, query, maxDepth=None, filter=None,
             queryFactory, getQueries = self._getQueryFactory(
                 query, queryFactory)
         try:
-            self.yieldRelationTokenChains(
+            six.next(self.yieldRelationTokenChains(
                 *self._parse(
                     query, maxDepth, filter, targetQuery,
                     targetFilter, getQueries) +
-                (False,)).next()
+                (False,)))
```
Comment (mgedmin) [inline src/zc/relation/catalog.py:1166]: Here also.

File: src/zc/relation/searchindex.py [hunk 4]
```diff
@@ -172,7 +173,7 @@ def _index(self, token, removals=None, remove=False):
                         indexed = self.index.get(rel)
                         if indexed is None:
                             iterator = reversed(stack)
-                            traversed = [iterator.next()]
+                            traversed = [six.next(iterator)]
```
Comment (mgedmin) [inline src/zc/relation/searchindex.py]: And here.

File: src/zc/relation/searchindex.py [hunk 5]
```diff
@@ -435,18 +438,18 @@ def getQueries(self, token, catalog, additions, removals, removed):
         for name in self.names:
             src = source[name]
             iterator = iter(src)
-            value = iterator.next() # should always have at least one
+            value = six.next(iterator)  # should always have at least one
```
Comment (mgedmin) [inline src/zc/relation/searchindex.py]: And here.

File: src/zc/relation/searchindex.py [hunk 6]
```diff
@@ -435,18 +438,18 @@ def getQueries(self, token, catalog, additions, removals, removed):
...
             vals.append([name, value, iterator, src])
         while 1:
             yield BTrees.family32.OO.Bucket(
                 [(name, value) for name, value, iterator, src in vals])
             for s in vals:
                 name, value, iterator, src = s
                 try:
-                    s[1] = iterator.next()
+                    s[1] = six.next(iterator)
```
Comment (mgedmin) [inline src/zc/relation/searchindex.py]: And here.

File: src/zc/relation/searchindex.py [hunk 7]
```diff
@@ -435,18 +438,18 @@ def getQueries(self, token, catalog, additions, removals, removed):
...
                 except StopIteration:
                     iterator = s[2] = iter(src)
-                    s[1] = iterator.next()
+                    s[1] = six.next(iterator)
```
Comment (mgedmin) [inline src/zc/relation/searchindex.py]: And here.

File: src/zc/relation/tokens.txt [hunk 8]
```diff
@@ -60,8 +60,97 @@ Now we can create some classes.  In the README example, the setup was a bit
...
+    ...         if not isinstance(other, Reference):
+    ...             raise ValueError('can only compare with Reference objects')
+    ...         return self._get_sorting_key() < other._get_sorting_key()
+    ...     def __eq__(self, other):
+    ...         # this doesn't work during conflict resolution.  See
+    ...         # zope.app.keyreference.persistent, 3.5 release, for current
+    ...         # best practice.
+    ...         if not isinstance(other, Reference):
+    ...             raise ValueError('can only compare with Reference objects')
+    ...         return self._get_sorting_key() == other._get_sorting_key()
```
Comment (mgedmin) [inline src/zc/relation/tokens.txt:104]: I don't like seeing only two of six rich comparison methods being defined. Can you add the @functools.total_ordering class decorator?

File: src/zc/relation/tokens.txt [hunk 9]
```diff
@@ -106,23 +195,26 @@ to sanely and transparently hold a set of persistent objects, try the
...
     ...         self.title = title
     ...         self.parts = BTrees.family32.IF.TreeSet()
     ...     # the next parts just make the tests prettier
     ...     def __repr__(self):
     ...         return '<Organization instance "' + self.title + '">'
-    ...     def __cmp__(self, other):
+    ...     def __lt__(self, other):
     ...         # pukes if other doesn't have name
-    ...         return cmp(self.title, other.title)
+    ...         return self.title < other.title
```
Comment (mgedmin) [inline src/zc/relation/tokens.txt:213]: I don't like seeing only one of six rich comparison methods being defined. If you use @functools.total_ordering, you can get away by defining two of six.

File: src/zc/relation/tokens.txt [hunk 10]
```diff
@@ -332,13 +469,18 @@ this were a real system.  The organization will be a direct object reference.
...
+    ...             self.principal_id,
+    ...             tuple(self.role_ids),
+    ...             self._organization.title,
+    ...         )
+    ...         _other = (
+    ...             other.principal_id,
+    ...             tuple(other.role_ids),
+    ...             other._organization.title,
+    ...         )
+    ...         return _self <_other
```
Comment (mgedmin) [inline src/zc/relation/tokens.txt:491]: I don't like seeing only one of six rich comparison methods being defined. If you use @functools.total_ordering, you can get away by defining two of six.

**Review by mgedmin — COMMENTED** [review]
LGTM, except for my hangup about rich comparison functions.

(Also add version classifiers to setup.py please?)

**Comment by ale-rt** [comment]
Thanks @mgedmin I think I did everything that was requested :)

File: src/zc/relation/README.txt [hunk 11]
```diff
@@ -109,6 +111,12 @@ name as the tokens.
     ...         return '<Employee instance "' + self.name + '">'
     ...     def __lt__(self, other):
     ...         return self.name < other.name
+    ...     def __eq__(self, other):
+    ...         return False
```
Comment (mgedmin) [inline src/zc/relation/README.txt]: Maybe stick with `self is other` like in the other two cases?

Comment (ale-rt) [reply src/zc/relation/README.txt]: Fixed, thanks!

**Review by mgedmin — APPROVED** [review]
LGTM

**Comment by icemac** [comment]
Is there still something necessary before this PR can be merged?

**Comment by ale-rt** [comment]
Not that I know :) It uses mr.developer to checkout zc.relationship for the tests but given that the two packages are deeply coupled this might even be a good thing.
After this is merged the branch should not be deleted otherwise https://github.com/zopefoundation/zc.relationship/pull/2 will fail

**Comment by icemac** [comment]
@ale-rt zopefoundation/zc.relationship#2 is now approved, too. Let's merge both PRs. (In zopefoundation repositories usually the creator of the PR merges his own PR to show that he has signed the contributor agreement.)

**Pull request merged**